From a63f6c00c5595084949006b98756fb3653d61d1e Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Tue, 30 Sep 2014 19:57:56 -0400
Subject: [drm] i915: don't crash if unable to setup stolen

Message-id: <1412107076-15369-1-git-send-email-rclark@redhat.com>
Patchwork-id: 96813
O-Subject: [RHEL7] drm/i915: don't crash if unable to setup stolen
Bugzilla: 1128939
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Rob Clark <robdclark@gmail.com>

RHBZ: 1128939
Upstream: sent.. although possibly this issue is only triggered on
          RHEL7 kernel due to missing early-reservation patches on
          arch/x86/kernel/early-quirks.c

On this baytrail NUC I would see:

  *ERROR* conflict detected with stolen region: [0xbb000000 - 0xbf000000]

Since stolen region is not setup, dev_priv->mm.stolen would not be
initialized, which results that vlv_pctx is not allocated in
valleyview_setup_pctx() which results in things going BOOM in
valleyview_check_pctx().

Things seem to work out much better if we check for NULL ptr first.

Signed-off-by: Rob Clark <robdclark@gmail.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/i915/intel_pm.c b/drivers/gpu/drm/i915/intel_pm.c
index ee72807..9d8de8b 100644
--- a/drivers/gpu/drm/i915/intel_pm.c
+++ b/drivers/gpu/drm/i915/intel_pm.c
@@ -3762,6 +3762,9 @@ static void valleyview_check_pctx(struct drm_i915_private *dev_priv)
 {
  unsigned long pctx_addr = I915_READ(VLV_PCBR) & ~4095;
 
+ if (!dev_priv->vlv_pctx)
+  return;
+
  WARN_ON(pctx_addr != dev_priv->mm.stolen_base +
         dev_priv->vlv_pctx->stolen->start);
 }
-- 
1.7.1